pkg/crawler/central/central.go

```diff
@@ -0,0 +1,376 @@
+
+// downloadIndex downloads the index file to a cached location and returns the path
+func (s *Source) downloadIndex() (err error) {
+	// Check if the file already exists in cache
+	if _, err := os.Stat(s.indexPath); err == nil {
+		s.logger.Info("Using cached Maven Central Index", slog.String("path", s.indexPath))
+		return nil
+	}
+
+	s.logger.Info("Downloading Maven Central Index", slog.String("url", s.url), slog.String("path", s.indexPath))
+
+	// Create the output file
+	outFile, err := os.Create(s.indexPath)
+	if err != nil {
+		return xerrors.Errorf("failed to create output file: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(s.indexPath)
+		}
+	}()
+	defer outFile.Close()
+
+	// Create the request
+	req, err := retryablehttp.NewRequest("GET", s.url, nil)
+	if err != nil {
+		return xerrors.Errorf("failed to create request: %w", err)
+	}
+
+	// Execute the request
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return xerrors.Errorf("failed to download index: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Check for successful response
+	if resp.StatusCode != http.StatusOK {
+		return xerrors.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	// Copy the response body to the output file
+	n, err := io.Copy(outFile, resp.Body)
+	if err != nil {
+		return xerrors.Errorf("failed to save index file: %w", err)
+	}
+
+	s.logger.Info("Downloaded Maven Central Index", slog.String("path", s.indexPath), slog.Int64("size", n))
+
+	return nil
+}
+
+func (s *Source) Read(ctx context.Context, recordCh chan<- types.Record) error {
+	s.logger.Info("Starting Maven Central Index processing")
+
+	// Download/Get the index file first
+	if err := s.downloadIndex(); err != nil {
+		return xerrors.Errorf("failed to get index file: %w", err)
+	}
+
+	// Make a queue to buffer records scanned from the index
+	records := make(chan data.Record, 64)
+
+	// Create a file-based reader
+	logger := log.New(io.Discard, "", log.LstdFlags)
+	reader := readers.NewChunk(logger, records, config.Index{
+		Mode: config.Mode{
+			Type: config.All,
+		},
+		Source: config.Source{
+			Type: config.Local, // Use local file source
+		},
+	}, s.indexPath, nil)
+
+	errCh := make(chan error)
+	defer close(errCh)
+
+	go func() {
+		defer close(records)
+		if err := reader.Read(); err != nil && !errors.Is(err, io.EOF) {
+			errCh <- xerrors.Errorf("failed to read central index: %w", err)
+		}
+	}()
+
+	// Collect GAVC records from central index
+	if err := s.read(ctx, records, errCh, recordCh); err != nil {
+		return xerrors.Errorf("failed to read central index: %w", err)
+	}
+
+	s.logger.Info("Records in central index", slog.Int("count", len(s.records)))
+
+	// Fetch SHA1s from Maven Central
+	if err := s.fetchSHA1s(ctx, recordCh); err != nil {
+		return xerrors.Errorf("failed to fetch SHA1s: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Source) read(ctx context.Context, records <-chan data.Record, errCh chan error, recordCh chan<- types.Record) error {
```

Comment:
It looks like we don't need this channel.
```suggestion
func (s *Source) read(ctx context.Context, records <-chan data.Record, errCh chan error) error {
```